File: kancore_py/myComponents/ImageButtonGraphics.py
from enum import Enum
from PyQt5.QtCore import Qt, pyqtSignal
from PyQt5.QtGui import QPixmap
from PyQt5.QtWidgets import QGraphicsPixmapItem
from myComponents.PicGraphic import PicGraphic


class ButtStyle(Enum):
    ButtNorm = 0,
    ButtPress = 1,
    ButtRelease = 2,
    ButtEnter = 3,
    ButtLeave = 4,


class ImageButtonGraphics(PicGraphic):
    leftclick = pyqtSignal()
    rightclick = pyqtSignal()
    buttonPicture = None
    pressPicture = None
    releasePicture = None
    enterPicture = None
    leavePicture = None
    buttonPictureOrigin = None
    pressPictureOrigin = None
    releasePictureOrigin = None
    enterPictureOrigin = None
    leavePictureOrigin = None
    isAnimateEnable = False
    currentButtStyle = None

    def __init__(self, buttNormPic, x, y, animeGeomOn, animeOpacOn, parent):
        super(ImageButtonGraphics, self).__init__(buttNormPic, x, y, animeGeomOn, animeOpacOn, parent)
        self.setAcceptHoverEvents(True)
        self.setAcceptDrops(True)
        self.buttonPicture = QPixmap(buttNormPic)
        self.pressPicture = QPixmap(buttNormPic)
        self.releasePicture = QPixmap(buttNormPic)
        self.enterPicture = QPixmap(buttNormPic)
        self.leavePicture = QPixmap(buttNormPic)
        self.currentButtStyle = ButtStyle.ButtNorm

    def setAllButtImage(self, buttNorm, buttEnter, buttPress):
        self.setButtonPicture(buttNorm)
        self.setEnterPicture(buttEnter)
        self.setLeavePicture(buttNorm)
        self.setPressPicture(buttPress)
        self.setReleasePicture(buttEnter)

    def setButtonPicture(self, pic):
        self.buttonPicture = pic

    def setEnterPicture(self, pic):
        self.enterPicture = pic

    def setLeavePicture(self, pic):
        self.leavePicture = pic

    def setPressPicture(self, pic):
        self.pressPicture = pic

    def setReleasePicture(self, pic):
        self.releasePicture = pic

    def mousePressEvent(self, event):
        if event.button() == Qt.LeftButton:
            self.setPixmap(self.pressPicture)
            self.currentButtStyle = ButtStyle.ButtPress

    def mouseReleaseEvent(self, event):
        if (event.button() & Qt.LeftButton) == 1:
            self.leftclick.emit()
            self.setPixmap(self.enterPicture)
            self.currentButtStyle = ButtStyle.ButtRelease
        elif (event.button() & Qt.RightButton) == 1:
            self.rightclick.emit()

    def hoverEnterEvent(self, event):
        QGraphicsPixmapItem.hoverEnterEvent(event)
        self.setPixmap(self.enterPicture)
        self.currentButtStyle = ButtStyle.ButtEnter

    def hoverLeaveEvent(self, event):
        QGraphicsPixmapItem.hoverLeaveEvent(event)
        self.setPixmap(self.leavePicture)
        self.currentButtStyle = ButtStyle.ButtLeave

File: kancore_cpp/myComponents/imagebuttongraphics.cpp
#include "imagebuttongraphics.h"

ImageButtonGraphics::ImageButtonGraphics(const QPixmap &buttNormPic, int x, int y, bool animeGeomOn, bool animeOpacOn,
                                         QGraphicsItem *parent)
    :PicGraphic(buttNormPic,x,y,animeGeomOn,animeOpacOn,parent)
{
    //使能hover事件
    this->setAcceptHoverEvents(true);
    this->setAcceptDrops(true);
    /*一开始所有按钮状态均为buttNormPic*/
    buttonPicture   = new QPixmap(buttNormPic);
    pressPicture    = new QPixmap(buttNormPic);
    releasePicture  = new QPixmap(buttNormPic);
    enterPicture    = new QPixmap(buttNormPic);
    leavePicture    = new QPixmap(buttNormPic);
    currentButtStyle=ButtNorm;

}

ImageButtonGraphics::~ImageButtonGraphics()
{

}

/**
 * @brief 一次性设置所有按钮图片,方便
 *
 * @param buttNorm
 * @param buttEnter
 * @param buttPress
 */
void ImageButtonGraphics::setAllButtImage(const QPixmap &buttNorm,const QPixmap &buttEnter,
                                          const QPixmap &buttPress)
{
    //设置按钮图片
    this->setButtonPicture(buttNorm);
    this->setEnterPicture(buttEnter);
    this->setLeavePicture(buttNorm);
    this->setPressPicture(buttPress);
    this->setReleasePicture(buttEnter);
}

/**
 * @brief 设置按钮图片,可以随时使用,可用于中途改变按键图片,灵活
 *
 * @param pic 输入的pximap按钮图片
 */
void ImageButtonGraphics::setButtonPicture(QPixmap pic)
{
    *buttonPicture=pic;
}
void ImageButtonGraphics::setEnterPicture(QPixmap pic)
{
    *enterPicture=pic;

}
void ImageButtonGraphics::setLeavePicture(QPixmap pic)
{
    *leavePicture=pic;

}
void  ImageButtonGraphics::setPressPicture(QPixmap pic)
{
    *pressPicture=pic;
}
void  ImageButtonGraphics::setReleasePicture(QPixmap pic)
{
    *releasePicture=pic;
}



/***********************************protected****************************************/
/**
 * @brief 鼠标事件
 *
 * @param e 鼠标事件类型
 */
void ImageButtonGraphics::mousePressEvent (QGraphicsSceneMouseEvent *event)
{
    if(event->button()==Qt::LeftButton)
    {
        this->setPixmap(*pressPicture);
        currentButtStyle=ButtPress;
    }
}
void ImageButtonGraphics::mouseReleaseEvent(QGraphicsSceneMouseEvent *event)
{
    //响应左击事件
    if((event->button()&Qt::LeftButton)==1)
    {
        emit leftclick();
        this->setPixmap(*enterPicture);
        currentButtStyle=ButtRelease;
    }
    //响应右击事件
    else if((event->button()&Qt::RightButton)==1)
        emit rightclick();
}
void ImageButtonGraphics::hoverEnterEvent(QGraphicsSceneHoverEvent *event)
{
    QGraphicsPixmapItem::hoverEnterEvent(event);
    this->setPixmap(*enterPicture);
    event=event;
    currentButtStyle=ButtEnter;
}
void ImageButtonGraphics::hoverLeaveEvent(QGraphicsSceneHoverEvent *event)
{
    QGraphicsPixmapItem::hoverLeaveEvent(event);
    this->setPixmap(*leavePicture);
    event=event;
    currentButtStyle=ButtLeave;
}

